.golangci.yml

```diff
@@ -0,0 +1,282 @@
+  modules-download-mode: readonly
+
+
+# output configuration options
+output:
+  # colored-line-number|line-number|json|tab|checkstyle|code-climate, default is "colored-line-number"
+  format: colored-line-number
+
+  # print lines of code with issue, default is true
+  print-issued-lines: true
+
+  # print linter name in the end of issue text, default is true
+  print-linter-name: true
+
+
+# all available settings of specific linters
+linters-settings:
+  govet:
+    # report about shadowed variables
+    check-shadowing: true
+  goimports:
+    # put imports beginning with prefix after 3rd-party packages;
+    # it's a comma-separated list of prefixes
+    local-prefixes: github.com/m3db/m3
+  gocyclo:
+    # minimal code complexity to report, 30 by default (but we recommend 10-20)
+    min-complexity: 10
+  maligned:
+    # print struct with more effective memory layout or not, false by default
+    suggest-new: true
+  dupl:
+    # tokens count to trigger issue, 150 by default
+    threshold: 100
+  goconst:
+    # minimal length of string constant, 3 by default
+    min-len: 3
+    # minimal occurrences count to trigger, 3 by default
+    min-occurrences: 3
+  depguard:
+    list-type: blacklist
+    include-go-root: false
+    packages:
+      - github.com/sirupsen/logrus
+      - github.com/golang/protobuf/jsonpb
+      - google.golang.org/protobuf/encoding/protojson
+      - github.com/golang/protobuf/proto
+      - google.golang.org/protobuf/proto
+    packages-with-error-messages:
+      # specify an error message to output when a blacklisted package is used
+      github.com/sirupsen/logrus: "logging is allowed only by logutils.Log"
+      github.com/golang/protobuf/jsonpb: "replace with github.com/gogo/protobuf/jsonpb"
+      google.golang.org/protobuf/encoding/protojson: "replace with github.com/gogo/protobuf/jsonpb"
+      github.com/golang/protobuf/proto: "replace with github.com/gogo/protobuf/proto"
+      google.golang.org/protobuf/proto: "replace with github.com/gogo/protobuf/proto"
+  misspell:
+    # Correct spellings using locale preferences for US or UK.
+    # Default is to use a neutral variety of English.
+    # Setting locale to US will correct the British spelling of 'colour' to 'color'.
+    locale: US
+    ignore-words:
+      - someword
+  lll:
+    # max line length, lines longer will be reported. Default is 120.
+    # '\t' is counted as 1 character by default, and can be changed with the tab-width option
+    line-length: 120
+    # tab width in spaces. Default to 1.
+    tab-width: 1
+  unused:
+    # treat code as a program (not a library) and report unused exported identifiers; default is false.
+    # XXX: if you enable this setting, unused will report a lot of false-positives in text editors:
+    # if it's called for subdir of a project it can't find funcs usages. All text editor integrations
+    # with golangci-lint call it on a directory with the changed file.
+    check-exported: false
```

Comment:
Thoughts on defaulting this to true? We probably shouldn't allow unused exported identifiers except in special situations (and generally we should be referencing things we're using, even if only in tests), especially if we're trying to be intentional about API surface area.

Reply:
does referencing the exported type from a test in the same package prevent the check from firing?

I think the false-positives from text editor integrations makes this a pain to enable.

Reply:
What I'd like to avoid is blindly exporting types that aren't actually being used to avoid API bloat, such that we catch exported types that can't be proven to be used in some capacity (or otherwise, `//nolint` them).

I use vanilla vim (I know, I'm a masochist), so I can't comment on workarounds for editors, but off the cuff I'd rather have our linting be defined by goal state rather than by issues/limitations with editor(s).

Reply:
@mway I agree that this would be a nice option in general.
But somewhat unexpetedly, after enabling it I'm getting messages on test functions everywhere:
```
(...)
src/aggregator/aggregator/flush_mgr_test.go:36:6: func `TestFlushManagerReset` is unused (unused)
src/dbnode/client/session_fetch_tagged_test.go:406:6: func `TestSessionFetchTaggedMergeWithRetriesTest` is unused (unused)
src/query/graphite/common/basic_functions_test.go:97:6: func `TestNormalize` is unused (unused)
(...)
```
Also, I believe we have a need to treat some pieces of this code base as a library, which would become complicated with this option. Leaving it at 'false` at least for now.

Reply:
Ah, yeah, I see what you mean. If we were able to skip test files for that specific linter, that seems like it would help, but if we're just going to get a bunch of noise because it doesn't understand that they're tests, agree that this isn't great.

I wonder whether it behaves differently in `_test` packages, which I think we should be using more broadly in general (although obviously not the case here in general, and probably wouldn't help us for the majority of the lint noise).